lib/diaspora_federation/entities/signed_retraction.rb

```diff
@@ -25,7 +25,48 @@ class SignedRetraction < Entity
       #   Contains a signature of the entity using the private key of the author of a post
       #   This signature is mandatory.
       #   @return [String] author signature
-      property :target_author_signature
+      property :target_author_signature, default: nil
+
+      # Generates XML and updates signatures
+      # @see Entity#to_xml
+      # @return [Nokogiri::XML::Element] root element containing properties as child elements
+      def to_xml
+        entity_xml.tap do |xml|
+          hash = to_h
+          self.class.update_signatures!(hash)
+
+          xml.at_xpath("target_author_signature").content = hash[:target_author_signature]
+        end
+      end
+
+      # Adds signature to a given hash with the key of the author
+      # if the signature is not in the hash yet and if the key is available.
+      #
+      # @param [Hash] data hash given for a signing
+      def self.update_signatures!(data)
+        if data[:target_author_signature].nil?
+          pkey = DiasporaFederation.callbacks.trigger(:fetch_private_key_by_id, data[:diaspora_id])
+          data[:target_author_signature] = Signing.sign_with_key(apply_signable_exceptions(data), pkey) unless pkey.nil?
+        end
+      end
+
+      # verifies the signature +target_author_signature+
+      # @param [Hash] data hash with data to verify
+      # @raise [SignatureVerificationFailed] if the signature is not valid or no public key is found
+      def self.verify_signatures(data)
+        pkey = DiasporaFederation.callbacks.trigger(:fetch_public_key_by_id, data[:diaspora_id])
+        raise SignatureVerificationFailed, "failed to fetch public key for #{data[:diaspora_id]}" if pkey.nil?
+        raise SignatureVerificationFailed, "wrong author_signature" unless Signing.verify_signature(
```

Comment:
I'd stop using a modifer-if/unless when you have to break it across several lines. Using the regular variant becomes much cleaner usually at this point.